DuckDuckGo/AutomationServer.swift

```diff
@@ -0,0 +1,340 @@
+        }
+        Task {
+            await self.executeScript(script, args: args, on: connection)
+        }
+    }
+
+    @MainActor
+    func getWindowHandle(on connection: NWConnection, url: URLComponents) {
+        let handle = self.main.currentTab
+        guard let handle else {
+            self.respondError(on: connection, error: "no window")
+            return
+        }
+        self.respond(on: connection, response: String(UInt(bitPattern: ObjectIdentifier(handle))))
+    }
+
+    @MainActor
+    func getWindowHandles(on connection: NWConnection, url: URLComponents) {
+        let handles = self.main.tabManager.model.tabs.map({ tab in
+            let tabView = self.main.tabManager.controller(for: tab)!
+            return String(UInt(bitPattern: ObjectIdentifier(tabView)))
+        })
+
+        if let jsonData = try? JSONEncoder().encode(handles),
+           let jsonString = String(data: jsonData, encoding: .utf8) {
+            self.respond(on: connection, response: jsonString)
+        } else {
+            // Handle JSON encoding failure
+            self.respondError(on: connection, error: "Failed to encode response")
+        }
+    }
+
+    @MainActor
+    func closeWindow(on connection: NWConnection, url: URLComponents) {
+        self.main.closeTab(self.main.currentTab!.tabModel)
+        self.respond(on: connection, response: "{\"success\":true}")
+    }
+
+    @MainActor
+    func switchToWindow(on connection: NWConnection, url: URLComponents) {
+        if let handleString = getQueryStringParameter(url: url, param: "handle") {
+            Logger.automationServer.info("Switch to window \(handleString)")
+            let tabToSelect: TabViewController? = nil
+            if let tabIndex = self.main.tabManager.model.tabs.firstIndex(where: { tab in
+                guard let tabView = self.main.tabManager.controller(for: tab) else {
+                    return false
+                }
+                return String(UInt(bitPattern: ObjectIdentifier(tabView))) == handleString
+            }) {
+                Logger.automationServer.info("found tab \(tabIndex)")
+                self.main.tabManager.select(tabAt: tabIndex)
+                self.respond(on: connection, response: "{\"success\":true}")
+            } else {
+                self.respondError(on: connection, error: "Invalid window handle")
+            }
+        } else {
+            self.respondError(on: connection, error: "Invalid window handle")
+        }
+    }
+
+    @MainActor
+    func newWindow(on connection: NWConnection, url: URLComponents) {
+        self.main.newTab()
+        let handle = self.main.tabManager.current(createIfNeeded: true)
+        guard let handle else {
+            self.respondError(on: connection, error: "no window")
+            return
+        }
+        // Response {handle: "", type: "tab"}
+        let response: [String: String] = ["handle": String(UInt(bitPattern: ObjectIdentifier(handle))), "type": "tab"]
```

Comment:
Do you have to use an ObjectIdentifier?  Tabs already have a UUID allocated them you could use if just need an id for looking up.

Reply:
I'd missed this, that's great thank you! So much simpler!